os/drivers/input/touchscreen.c

```diff
@@ -183,23 +196,60 @@ static int touch_close(FAR struct file *filep)
 
 static ssize_t touch_read(FAR struct file *filep, FAR char *buffer, size_t buflen)
 {
-	FAR struct touchscreen_s *priv;
-	size_t outlen;
-	int ret;
+	FAR struct touchscreen_s *dev;
+	struct touch_point_buffer_s *tp_buf;
 
-	priv = filep->f_inode->i_private;
-
-	if (!priv || buflen < 1) {
+	ssize_t recvd = 0;
+	int16_t tail;
+	int ret;
+	
+	dev = filep->f_inode->i_private;
+	if (!dev || buflen < sizeof(struct touch_point_s)) {
 		return -EINVAL;
 	}
+
+	DEBUGASSERT(dev->tp_buf.buffer);
+
+	tp_buf = &dev->tp_buf;
+
 	/* Wait for semaphore to prevent concurrent reads */
-	touch_semtake(&priv->sem, false);
+	touch_semtake(&tp_buf->sem, false);
```

Comment:
you should consider below.
1. Enable / disable condition (open -> ioctl(TSIOC_DISABLE) -> read case)
2. if user increased size of ringbuffer and it fully filled with touch event, but app request stop in the middle of  read event. 
3. How app can realize proper buflen? (actually this also violate posix, it should return even 1byte.)
4. In posix, write -> read then written value must be readed. Current code violate posix
5. Someday if we apply micro kernel fully, because of differrent address logic in app & kernel, this will not work properly because address of buffer can't be relalized in kernel.

Reply:
Thank you for review.

we need to remove ioctl(TSIOC_DISABLE), I think there is nothing for the application to do in order to read touch data.

Could you please explain more about `In posix, write -> read then written value must be readed. Current code violate posix`?

we are using write read in serial.c also.
is the problem related to no code in write?

Reply:
TSIOC_DISABLE is added for suspend case as was asked that interrupt needs to be disabled. 
App need to ensure to call TSIOC_DISABLE and TSIOC_ENABLE when going to suspend and resume. doesn't it?